clang/include/clang/3C/ConstraintsGraph.h

```diff
@@ -235,6 +239,32 @@ class ConstraintsGraph : public DataGraph<Atom *> {
   // be able to retrieve them from the graph.
   std::set<ConstAtom *> &getAllConstAtoms();
 
+  typedef DataEdge<Atom*> EdgeType;
+
+  bool getNeighbors(Atom *A, std::vector<EdgeType*> &DataSet, bool Succ,
```

Comment:
Almost. This is part of a minimal-change hack to avoid rewriting all the other code in this file. `EdgeType` exists elsewhere as a template parameter, which can't be accessed outside the class. This `getNeighbors` fills `DataSet` with actual graph edges, while the other just fills it with data from the target nodes. But that data lacks the new constraints, because this graph class wasn't expected to store different data in the edges. So we either need a graph class rewrite or this overloaded function. I'd prefer having the former, but creating it seems like a waste of time until we use it more broadly.

I'll add some comments, but let me know if I should document this more elsewhere, open issues, or whatever else.

Reply:
This should work to get rid of the duplicated code:
https://github.com/correctcomputation/checkedc-clang/compare/add_constraints...correctcomputation:add_constraints.reduce_duplicated_graph_code